sympy/stats/stochastic_process_types.py

```diff
@@ -759,3 +765,204 @@ def limiting_distribution(self):
         eqs.append(sum(wi) - 1)
         soln = list(linsolve(eqs, wi))[0]
         return ImmutableMatrix([[sol for sol in soln]])
+
+
+class BernoulliProcess(DiscreteTimeStochasticProcess):
+    """
+    The Bernoulli process consists of repeated
+    independent Bernoulli trials with the same parameter `p`.
+    It's assumed that the probability `p` applies to every
+    trial and that the outcomes of each trial
+    are independent of all the rest. Therefore Bernoulli Processs
+    is Discrete State and Discrete Time Stochastic Process.
+
+    Parameters
+    ==========
+
+    sym: Symbol/string_types
+    success: Integer/string_types
+            The event which is considered to be success, by default is 1.
+    failure: Integer/string_types
+            The event which is considered to be failure, by default is 0.
+    p: Real Number between 0 and 1
+            Represents the probability of getting success.
+
+    Examples
+    ========
+
+    >>> from sympy.stats import BernoulliProcess, P, E
+    >>> from sympy import Eq, Gt, Lt
+    >>> B = BernoulliProcess("B", p=0.7, success=1, failure=0)
+    >>> B.state_space
+    FiniteSet(0, 1)
+    >>> (B.p).round(2)
+    0.70
+    >>> B.success
+    1
+    >>> B.failure
+    0
+    >>> X = B[1] + B[2] + B[3]
+    >>> P(Eq(X, 0)).round(2)
+    0.03
+    >>> P(Eq(X, 2)).round(2)
+    0.44
+    >>> P(Eq(X, 4)).round(2)
+    0
+    >>> P(Gt(X, 1)).round(2)
+    0.78
+    >>> P(Eq(B[1], 0) & Eq(B[2], 1) & Eq(B[3], 0) & Eq(B[4], 1)).round(2)
+    0.04
+    >>> B.joint_distribution(B[1], B[2])
+    JointDistributionHandmade(Lambda((B[1], B[2]), Piecewise((0.7, Eq(B[1], 1)),
+    (0.3, Eq(B[1], 0)), (0, True))*Piecewise((0.7, Eq(B[2], 1)), (0.3, Eq(B[2], 0)),
+    (0, True))))
+    >>> E(2*B[1] + B[2]).round(2)
+    2.10
+    >>> P(B[1] < 1).round(2)
+    0.30
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Bernoulli_process
+    .. [2] https://mathcs.clarku.edu/~djoyce/ma217/bernoulli.pdf
+
+    """
+
+    index_set = S.Naturals0
+
+    @property
+    def p(self):
+        return self.args[3]
+
+    @property
+    def distribution(self):
+        return self.args[2]
+
+    @property
+    def success(self):
+        return self.args[4]
+
+    @property
+    def failure(self):
+        return self.args[5]
+
+    def __new__(cls, sym, p, success=1, failure=0):
+        _value_check(p >= 0 and p <= 1, 'Value of p must be between 0 and 1.')
+        p = _sympify(p)
+        sym = _symbol_converter(sym)
+        state_space = _set_converter([success, failure])
+        return Basic.__new__(cls, sym, state_space, BernoulliDistribution(p), p,
+                             sympify(success), sympify(failure))
```

Comment:
`_sympify`

Reply:
Actually using `_sympify` there produces sympify error when we pass `success="H"` and `failure="T"`

Reply:
Okay, in `SingleFiniteDistribution` too `sympify` is used, so we can go ahead with this.

Reply:
It should be changed to `_sympify` basically everywhere #18066.

What test is failing if it is changed? Is the test valid?

Reply:
I think that @Smit-create a different PR for introducing this change. This one can wait till that is merged.

Reply:
```
>>>_sympify("H")
SympifyError                              Traceback (most recent call last)
<ipython-input-6-5bb9a00aa23c> in <module>
----> 1 _sympify("H")

~\Anaconda3\lib\site-packages\sympy\core\sympify.py in _sympify(a)
    416 
    417     """
--> 418     return sympify(a, strict=True)
    419 
    420 

~\Anaconda3\lib\site-packages\sympy\core\sympify.py in sympify(a, locals, convert_xor, strict, rational, evaluate)
    338 
    339     if strict:
--> 340         raise SympifyError(a)
    341 
    342     if iterable(a):

SympifyError: SympifyError: 'H'

```

Reply:
> I think that @Smit-create a different PR for introducing this change. This one can wait till that is merged.

So should I change sympify to _sympify() everywhere in the module and open a PR?

Reply:
Probably Yes.

Reply:
But facing the above error how it can be resolved?

Reply:
That's the point of `_sympify`: it doesn't parse/eval strings.

My question is: does the stats code expect to be able to pass strings? If so then why?

Reply:
Yes, in stats module the events can be passed as strings, for example in `BernoulliDistribution`, `"Heads"` or `"Tails"` can be passed as strings. Though I am not sure about the need of sympifying these strings. 
What will happen if we don't sympify them, this needs to be answered which will need some experimentation.

Reply:
> Yes, in stats module the events can be passed as strings, for example in `BernoulliDistribution`, `"Heads"` or `"Tails"` can be passed as strings. Though I am not sure about the need of sympifying these strings.
> What will happen if we don't sympify them, this needs to be answered which will need some experimentation.

With out sympify it will produce an` Assertion error` in `_test_args(BernoulliProcess("B", p=0.5, success=1,failure=0)`

Reply:
I found that the only need  of sympify here could be to pass the `_test_args`

Reply:
A better approach across stats would be to do:
```python
def _sym_sympify(arg):
    if isinstance(arg, str):
        return Symbol(arg)
    else:
        return _sympify(arg)
```

Reply:
```
 def __new__(cls, sym, success, failure, p):
        _value_check(p >= 0 and p <= 1, 'Value of p must be between 0 and 1.')
        sym = _symbol_converter(sym)
        state_space = _set_converter([success, failure])
        if S(success).is_integer and S(failure).is_integer:
            success = S(success)
            failure = S(failure)
        else:
            success = Symbol(success)
            failure = Symbol(failure)
        return Basic.__new__(cls, sym, state_space, BernoulliDistribution(p), S(p), success,failure)

```
I found a way to eliminate the use of sympify, and pass the tests too, should I go for it?

Reply:
May be, `sympy.stats.rv._symbol_converter` can be of help here.
@Smit-create Can you try the following?
```python
_sympify(_symbol_converter(arg))
```

Reply:
> ```python
> _sympify(_symbol_converter(arg))
> ```

This produces error while `success=1` and `failure=0`

Reply:
> A better approach across stats would be to do:
> 
> ```python
> def _sym_sympify(arg):
>     if isinstance(arg, str):
>         return Symbol(arg)
>     else:
>         return _sympify(arg)
> ```

I think this could be used

Reply:
> A better approach across stats would be to do:
> 
> ```python
> def _sym_sympify(arg):
>     if isinstance(arg, str):
>         return Symbol(arg)
>     else:
>         return _sympify(arg)
> ```

Should I add this in stats.rv.py so that I could be used where-ever required. And after merging this open another PR for making the required change in other stats files. So should I go for this?